### PR TITLE
Ebarriosjr/creating gcp projects plugin

### LINKED ISSUE #626: [Plugin] Create new Google Cloud Platform Project
## Summary

Depends on https://github.com/spotify/backstage/issues/289

This plugin would expose a simple form that lets users create a new GCP Project. Example from inside Spotify:

Screenshot 2020-04-23 at 11 30 08

Additional settings could/should be available as well. 

## Project website (if applicable)

https://cloud.google.com/resource-manager/docs/creating-managing-projects

## Context

Having project creation in Backstage allows you to streamline how projects are set up in your organisation. This could include adding custom parameters such as billing information, labels, etc.

### PR DESCRIPTION
**Pull request by ebarriosjr**
## Hey, I just made a Pull Request!

I added an implementation fo the gcp-projects plugin. It is not perfect and may need some work but it should be a good stepping stone. 
Gihub Issue: [Plugin] Create new Google Cloud Platform Project #626

#### :heavy_check_mark: Checklist

- [ ] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [ ] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

**Comment by stefanalund** [comment]
@ebarriosjr thanks! Could you please add some screenshots?

File: plugins/gcp-projects/src/api/GCPClient.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    };
+
+    const newProject: Project = {
+      name: projectName,
+      projectId: projectId,
+    };
+
+    const body = JSON.stringify(newProject);
+
+    const url = `https://content-cloudresourcemanager.googleapis.com/v1/projects`;
```
Comment (stefanalund) [inline plugins/gcp-projects/src/api/GCPClient.ts:105]: This URL is specified in multiple places. I would suggest creating a `baseURL` variable that can be reused across the code.

Comment (ebarriosjr) [reply plugins/gcp-projects/src/api/GCPClient.ts]: Done :)

**Comment by ebarriosjr** [comment]
> @ebarriosjr thanks! Could you please add some screenshots?

Working on this. I have to create some mock data.

**Comment by ebarriosjr** [comment]
Hi @stefanalund here are some screenshots:
listpage-gcp
newProject-gcp
confirm-project

**Comment by stefanalund** [comment]
@katz95 can you please review this from a UX perspective?

**Comment by katz95** [comment]
Hi @ebarriosjr ! Thanks for sending this in - this is a great start! To ensure that your plugin is ready for success, here are a couple of UX requests:

1. Please add a header to all pages of the experience! You can check out the different types of headers in our Storybook [here](http://storybook.backstage.io/?path=/story/header--tool).

2. Please place the button for creating a GCP project on the upper right hand side of the plugin page (under the header). Here's an example. We also typically have a support button in that corner as well. Attached is a template view of what I'm talking about:
Examples

3. If you could format the page in the 3rd screenshot similarly to the following screenshot, that would be great.
Example2

You can also view more in depth design specs [here](https://www.figma.com/community/file/850673348101741100/Backstage-Design-System), at our Figma community Backstage file (check under the SimpleStepper page on the left hand side).
Screenshot 2020-07-14 at 10 09 51 PM

Let me know if you have any questions! Thanks again!

**Comment by stefanalund** [comment]
We are excited to land this. Let us know if you have any questions about the UX feedback.

**Comment by ebarriosjr** [comment]
New Screenshots:
listpage-gcp
detailspage-gcp
newProject-gcp
newProject2-gcp

**Comment by ebarriosjr** [comment]
Hi @katz95 
Could you please let me know if the new screenshots look ok to you? 
Thanks!

File: plugins/gcp-projects/src/components/ProjectDetailsPage/ProjectDetailsPage.tsx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+      <Typography variant="h6" color="error">
+        Failed to load build, {status.error.message}
+      </Typography>
+    );
+  }
+
+  const details = status.value;
+
+  return (
+    <TableContainer component={Paper} className={classes.table}>
```
Comment (stefanalund) [inline plugins/gcp-projects/src/components/ProjectDetailsPage/ProjectDetailsPage.tsx:89]: You could consider using the Backstage `Table` component instead: http://storybook.backstage.io/?path=/story/table--default-table

File: plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+  if (error) {
+    return (
+      <Typography variant="h2" color="error">
+        {error.message}{' '}
+      </Typography>
+    );
+  }
+
+  return (
+    <TableContainer component={Paper}>
```
Comment (stefanalund) [inline plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx:92]: Same comment as above. Using the default Table component saves code and ensures a more consistent experience across different plugins.

**Review by stefanalund — COMMENTED** [review]
Very close to being shippable 💯

File: plugins/gcp-projects/package.json [hunk 4]
```diff
@@ -0,0 +1,47 @@
+{
+  "name": "@backstage/plugin-gcp-projects",
+  "version": "0.1.1-alpha.16",
```
Comment (Rugvip) [inline plugins/gcp-projects/package.json:3]: Need to update these versions to the latest before we merge. This and the deps below.

**Review by Rugvip — COMMENTED** [review]
Awesome stuff, thanks! :grin:

File: plugins/gcp-projects/src/components/NewProjectPage/NewProjectPage.tsx [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+export const NewProjectPage = () => {
+  return (
+    <Page theme={pageTheme.service}>
+      <Header title="New GCP Project" type="tool">
+        {labels}
+      </Header>
+      <Content>
+        <ContentHeader title="">
+          <SupportButton>Support Button</SupportButton>
```
[the remark this answers is not in the document]
Comment (ebarriosjr) [reply plugins/gcp-projects/src/components/NewProjectPage/NewProjectPage.tsx]: Added text inside the support button.

File: plugins/gcp-projects/src/components/ProjectDetailsPage/ProjectDetailsPage.tsx (hunk 2 above)
Comment (ebarriosjr) [reply plugins/gcp-projects/src/components/ProjectDetailsPage/ProjectDetailsPage.tsx:89]: Hi @stefanalund sadly the link is broken. Do you know if it changed?

File: plugins/gcp-projects/package.json (hunk 4 above)
Comment (ebarriosjr) [reply plugins/gcp-projects/package.json]: Updated dependencies :)

File: plugins/gcp-projects/src/components/ProjectDetailsPage/ProjectDetailsPage.tsx (hunk 2 above)
Comment (marcuseide) [reply plugins/gcp-projects/src/components/ProjectDetailsPage/ProjectDetailsPage.tsx:89]: Correct link: https://backstage.io/storybook/?path=/story/table--default-table

**Review by marcuseide — COMMENTED** [review]
Hey @ebarriosjr , looks great! If you want to use that `<Table />` component I think we can merge after that.

**Comment by freben** [comment]
@ebarriosjr Hi! If I understand things correctly, the current status is that a small conflict remains plus a desire to change table implementation. Is there something we can help out with?

**Comment by ebarriosjr** [comment]
Hi @freben, I have been busy with work and haven't been able to solve the conflict and do the table change. But I will as soon as I get a sec.

**Comment by ebarriosjr** [comment]
@freben should be good to go now... any feedback will be appreciated.

File: plugins/gcp-projects/src/plugin.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import { ProjectListPage } from './components/ProjectListPage';
+import { ProjectDetailsPage } from './components/ProjectDetailsPage';
+import { NewProjectPage } from './components/NewProjectPage';
+
+export const rootRouteRef = createRouteRef({
+  path: '/gcp-projects',
+  title: 'GCP Projects',
+});
+export const ProjectRouteRef = createRouteRef({
+  path: '/gcp-projects/project',
```
Comment (freben) [inline plugins/gcp-projects/src/plugin.ts:27]: this should be using a named route segment instead, `/gcp-projects/project/:id`

File: plugins/gcp-projects/src/plugin.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  title: 'GCP Project Page',
+});
+export const NewProjectRouteRef = createRouteRef({
+  path: '/gcp-projects/new',
+  title: 'GCP Project Page',
```
Comment (freben) [inline plugins/gcp-projects/src/plugin.ts:32]: change to `Create a GCP Project`

File: plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+          ))}
+        </TableBody>
+      </Table>
+    </Table>
+  );
+};
+
+export const ProjectListPage = () => {
+  return (
+    <Page theme={pageTheme.service}>
```
Comment (freben) [inline plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx:149]: probably the tool theme?

File: plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+      <Header title="GCP Projects" type="tool">
+        {labels}
+      </Header>
+      <Content>
+        <ContentHeader title="">
+          <Button variant="contained" color="primary" href="/gcp-projects/new">
```
Comment (freben) [inline plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx:155]: i wonder if we should make these like `component={NavLink} to="new"` to not cause page reloads

File: plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+            New Project
+          </Button>
+          <SupportButton>All your software catalog entities</SupportButton>
```
Comment (freben) [inline plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx:158]: Need to change the text here

File: plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+  });
+
+  if (loading) {
+    return <LinearProgress />;
+  }
+
+  if (error) {
+    return (
+      <Typography variant="h2" color="error">
+        {error.message}{' '}
```
Comment (freben) [inline plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx:85]: extra space

File: plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    return projects;
+  });
+
+  if (loading) {
+    return <LinearProgress />;
+  }
+
+  if (error) {
+    return (
+      <Typography variant="h2" color="error">
```
Comment (freben) [inline plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx:84]: there's an `<Alert severity="error">` thing to use

File: plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    const token = await googleApi.getAccessToken(
+      'https://www.googleapis.com/auth/cloud-platform.read-only',
+    );
+
+    const projects = api.listProjects({ token });
+    return projects;
+  });
+
+  if (loading) {
+    return <LinearProgress />;
```
Comment (freben) [inline plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx:79]: should use `<Progress>` in backstage core that has some additional features on top of the default progress

File: plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+  Typography,
+  Button,
+} from '@material-ui/core';
+import React from 'react';
+import { useAsync } from 'react-use';
+import { GCPApiRef, Project } from '../../api';
+
+const LongText = ({ text, max }: { text: string; max: number }) => {
+  if (text.length < max) {
+    return <span>{text}</span>;
```
Comment (freben) [inline plugins/gcp-projects/src/components/ProjectListPage/ProjectListPage.tsx:49]: can just skip the span here and below

**Review by freben — APPROVED** [review]
Hey! I had a look-through and had a bunch of little nits - do not worry, I think we should merge this and then we can make any additional tweaks later in a separate PR and I added them mainly as notes for the future and for myself. Thank you very much for your work!

File: plugins/gcp-projects/src/api/GCPApi.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+});
+
+export type GCPApi = {
+  listProjects: ({ token }: { token: string }) => Promise<Project[]>;
+  getProject: (projectId: string, token: Promise<string>) => Promise<Project>;
+  createProject: (
+    projectName: string,
+    projectId: string,
+    owner: string,
+    token: string,
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply plugins/gcp-projects/src/api/GCPApi.ts:32]: Gonna throw in that I don't think token should be of the signature either, and instead be grabbed from some source that is passed to the implementation.

**Comment by stefanalund** [comment]
@ebarriosjr do you want to go ahead and add this plugin to the Marketplace? https://backstage.io/plugins

**Pull request merged**